tsk/util/Bitlocker/BitlockerParser.cpp

```diff
@@ -0,0 +1,1553 @@
+}
+
+#define BITLOCKER_DIFFUSER_ROTATE_LEFT(a,n)  (((a) << (n)) | ((a) >> ((sizeof(a) * 8)-(n))))
+
+/**
+* Decrypt data using diffuser A
+*
+* @param data     Data to decrypt
+* @param dataLen  Length of data (in bytes)
+* @param result   Result buffer
+*/
+void BitlockerParser::decryptDiffuserA(uint8_t* data, uint16_t dataLen, uint8_t* result) {
+
+    // Copy the data into the result buffer
+    memcpy(result, data, dataLen);
+
+    // Set up a 4-byte version of the result buffer
+    uint32_t* result32 = (uint32_t*)result;
+    uint16_t result32len = dataLen / 4;
+
+    uint16_t shiftBits[] = { 9, 0, 13, 0 };
+    for (int cycle = 0; cycle < 5; cycle++) {  // Five cycles
+        for (int index = 0; index < result32len; index++) {
+
+            int indexMinusTwo = (index - 2 + result32len) % result32len;  // Add result32len to prevent negative result
+            int indexMinusFive = (index - 5 + result32len) % result32len;
+            result32[index] = result32[index] + 
+                (result32[indexMinusTwo] ^ BITLOCKER_DIFFUSER_ROTATE_LEFT(result32[indexMinusFive], shiftBits[index % 4]));
+        }
+    }
+}
+
+/**
+* Decrypt data using diffuser B
+* 
+* @param data     Data to decrypt
+* @param dataLen  Length of data (in bytes)
+* @param result   Result buffer
+*/
+void BitlockerParser::decryptDiffuserB(uint8_t* data, uint16_t dataLen, uint8_t* result) {
+
+    // Copy the data into the result buffer
+    memcpy(result, data, dataLen);
+
+    // Set up a 4-byte version of the result buffer
+    uint32_t* result32 = (uint32_t*)result;
+    uint16_t result32len = dataLen / 4;
+
+    uint16_t shiftBits[] = { 0, 10, 0, 25 };
+    for (int cycle = 0; cycle < 3; cycle++) {  // Three cycles
+        for (int index = 0; index < result32len; index++) {
+
+            int indexPlusTwo = (index + 2) % result32len;
+            int indexPlusFive = (index + 5) % result32len;
+            result32[index] = result32[index] +
+                (result32[indexPlusTwo] ^ BITLOCKER_DIFFUSER_ROTATE_LEFT(result32[indexPlusFive], shiftBits[index % 4]));
+        }
+    }
+}
+
+/**
+* Decrypt the data that was read from the given offset using AES-CBC with elephant diffuser (128 or 256 bit)
+*
+* @volumeOffset Offset to the data relative to the start of the volume. Expected to be sector-aligned.
+* @data         Data to decrypt. Will hold the decrypted data.
+*
+* @return 0 on success, -1 on error.
+*/
+int BitlockerParser::decryptSectorAESCBC_diffuser(uint64_t offset, uint8_t* data) {
+
+    // Make temporary buffer to copy encrypted data to
+    uint8_t* encryptedData = (uint8_t*)tsk_malloc(m_sectorSize);
+    uint8_t* tempBuffer = (uint8_t*)tsk_malloc(m_sectorSize);
+    if (encryptedData == nullptr || tempBuffer == nullptr) {
+        writeError("BitlockerParser::decryptSectorAESCBC_noDiffuser: Error allocating encryptedData or tempBuffer");
+        return -1;
+    }
+    memcpy(encryptedData, data, m_sectorSize);
+    memset(data, 0, m_sectorSize);
```

Comment:
same comment as before, should we be passing in size of data?

Reply:
Can be resolved based on previous comment https://github.com/sleuthkit/sleuthkit/pull/2923#discussion_r1610131510